interp/iostream_test.go

```diff
@@ -0,0 +1,321 @@
+package interp
```

Comment:
I started leaving individual comments on this file, but I'm not sure I agree with the approach you're taking here. These are white-box tests that test internal implementation details. And as noted below, I think some things don't really needs tests at all, like the double-close error.

It seems to me we can test most, if not all, of the expected behaviours by running it through the public `interp` API like the rest of the interpreter tests.

Can we make these regular `interpTests` tests? For example:

```go
var interpTests = []interpTest{
	// ...
	{`BEGIN { print system("kill -1 $$") }`, "", "257\n", "", ""},
}
```

Many of these should even work on Windows, as the default shell there is `sh` too. If not, you can use a special `# !windows-gawk` or `# !windows-goawk` comment -- see [exclude logic](https://github.com/benhoyt/goawk/blob/f2d5f1f1f831a193901126379d9401949d5f6185/interp/interp_test.go#L1005), and we could add another rule so just `# !windows` works too.

There's also the separate [TestShellCommand](https://github.com/benhoyt/goawk/blob/f2d5f1f1f831a193901126379d9401949d5f6185/interp/interp_test.go#L1440) if that's useful to look at.

So overall, let's see how far we can get making these black-box interp tests similar to the existing ones.

Reply:
Ok. I can move as much as possible into `interp_test.go`. But I think the lower-level I/O tests (i.e. "DoubleClose") are still important. I commented on that elsewhere.

I think having separate lower level tests are important for regression. If you have a bug appear it is better to isolate it at as low a level as possible. If a bug occurs in the `iostream` later, it will manifest in the higher level `interpTests` as well as in the stream tests. If the bug occurs at a higher level above the `iostream` it will only manifest in the `interpTests`. This makes it easier to diagnose.

I think your distinction between white box is mostly dependent on `package` boundaries. I am trying to unit test I/O in streams without worrying about the rest of `interp`. This is an attempt to create stratification. Maybe if I moved `iostreams.go` to a separate package it wouldn't be so bothersome to unit test them in isolation?

Reply:
See my comment above as well, but in short: I'm okay with a couple of "programmer error" tests that test implementation details, but yeah, let's move the rest to interp tests.

I think the occasional programmer error test is okay to be white box, but in general I disagree somewhat with you here: I think testing top-level contracts is much more useful and powerful than testing implementation details. Doing so allows you to refactor the internals majorly and have your test suite just keep on telling you what you need to know. You're right that sometimes lower-level tests allow you do diagnose a bug more quickly (closer to its source), but that's fairly rare.

I don't think it has much to do with package boundaries: what I'm suggesting is we test (as much as possible) the contract of essentially the entire program against the spec (POSIX spec or Gawk or whatever). That's what has allowed me to run these tests against multiple implementations (GoAWK, Gawk, nawk).

Reply:
I haven't had much time this week but I was working out where to put the tests in interp_test.

Reply:
No problem. I was thinking it'd be good if as many of them as possible could go in the list of generic interp tests, see the [`interp_tests`](https://github.com/benhoyt/goawk/blob/f2d5f1f1f831a193901126379d9401949d5f6185/interp/interp_test.go#L43) slice.

Reply:
I've moved everything other than the test for double-close conditions to `interp_test.go`. This was a problem for `outCmdStream`-s because they did not detect the double close as error and overwrote the `exitCode`.

Reply:
Also, I really don't think Windows supports POSIX signals but you're welcome to try if you have a Windows box handy. IIRC, cygwin offers some sort of emulation but I wouldn't assume go on Windows is compatible with this.